### PR TITLE
Add RISC-V native-code backend

### PR DESCRIPTION
**Pull request by nojb**
This PR proposes the addition of a backend for the native-code compiler targeting the RV64G RISC-V architecture.

## Background

This backend has been maintained out-of-tree for several years. It is already being used to build the OCaml packages of the Fedora distribution on RISC-V hardware (cc @rwmjones). Several bugs were found and fixed this way.

If this PR is accepted, I commit to maintaining it for the foreseeable future. I expect the maintenance burden to be rather low (as for the rest of the native-code backends). Most of the effort in the last couple of years (apart from the occasional bug fix) was related to it being maintained out-of-tree.

## Testing and CI integration

Recently @avsm made a [HiFive board](https://www.sifive.com/boards/hifive-unleashed) available to help upstream this work. The board is sitting in the offices of INRIA Paris and is connected to INRIA's CI system (thanks @shindere!). However, it suffers from some stability and performance issues and the setup is not yet ready to be used in production. So there is some work that remains to be done in that department, and it will not take place in the short term due to the COVID lockdown.

In spite of it, I am submitting the PR to increase visibility, and to invite anyone who wishes, to comment on it. Instructions for a local testing setup using a RISC-V/QEMU VM are available below.

## Setting up a VM for testing locally

A docker container with Fedora running on a RISC-V/QEMU VM is [available](https://hub.docker.com/r/sorear/fedora-riscv-wip). It is a little outdated, but it is easy to install and enough to build the compiler and run the testsuite.

The following instructions were tested under Ubuntu 18.04.

```
$ docker run --privileged -it sorear/fedora-riscv-wip
```

Once logged into the container, you need to install the packages `git` and `make`:
```
# dnf install -y git make
```

Lastly, you can clone the repository, and build the compiler:

```
# git clone -b riscv https://github.com/nojb/ocaml
# cd ocaml
# ./configure && make
```

(in my machine, it took ~30m to build the compiler with `-j4` + testsuite with `parallel`). 

## Reference material

- Instruction set specification: https://riscv.org/specifications/isa-spec-pdf/
  See also https://rv8.io/isa
- ELF ABI specification: https://github.com/riscv/riscv-elf-psabi-doc/blob/master/riscv-elf.md
- Assembly language manual: https://github.com/riscv/riscv-asm-manual/blob/master/riscv-asm.md
  See also https://rv8.io/asm

## License information

This work was done before I joined my current employer @LexiFi and so I may need to sign a separate copyright assignment, as my OCaml contributions currently fall under LexiFi's copyright.

## Call for review

The backend needs to be reviewed. I am cc-ing those who I know are familiar with the native-code backend, but of course any and all comments are welcome.

cc @xavierleroy @alainfrisch @mshinwell @rwmjones @stedolan @gretay-js @kayceesrk 

## Possible future enhancements

- 32-bit support
- Other variants of the RISC-V architecture (compact encodings, soft float, etc)

## Technical information

Writing a new backend for `ocamlopt` involves roughly three different pieces: describing the target processor, implementing a code generator, and the runtime assembly glue. Below are some comments about each part. They are meant to highlight some specific features of the backend, not meant to be exhaustive.

The backend does not currently support Spacetime nor function sections (#8526).

### Register assignment

The description of the target processor (`asmcomp/riscv/proc.ml`) specifies (among other things) which registers are used to pass arguments to OCaml function calls, and which registers are reserved to cache important variables from the runtime (such as the current heap allocation pointer or exception handler). Additionally every register is assigned an integer label used to identify it. This is done in `asmcomp/riscv/proc.ml`. For RISC-V, the following mapping is used:

| Register | Number | Purpose |
| :-------------: | :-------------: | ---- | 
| `zero`  | | always zero  |
| `ra` | | return address |
| `sp` `gp` `tp`| | | stack pointer, global pointer, thread pointer |
| `a0-a7` | 0-7 | arguments/results | 
| `s2-s9` | 8-15 | arguments/results | 
| `t2-t6` | 16-20 | temporary | 
| `t0-t1` | 21-22 | temporary (used by code generator) | 
| `s0` | 23 | domain pointer (preserved by C) | 
| `s1` | 24 | trap pointer (preserved by C) | 
| `s10` | 25 | allocation pointer (preserved by C) | 
| `s11` | 26 | allocation limit (preserved by C) | 

The register names are the official ABI names. The "s" registers are preserved by C function calls (callee-save) in the standard calling convention.

Registers 0-20 are available for the register allocator. The rest are either reserved for specific purposes, or used by code generator and so not available for register allocation.

Registers 16-20 (`t2-t6`) can be used by the register allocator but are also used by the dynamic
linker (PLT stubs), so they cannot live across an allocation (which may call a C function).

### Code generator

The purpose of the code generator (`asmcomp/riscv/emit.mlp`) is to translate the `Linear` language into textual assembly code, and to customize the `Cmm -> Mach` translation (`asmcomp/riscv/selection.ml`), where needed to account for specificities of the target architecture.

For RISC-V, the following adjustments are made:

- Recognize the `fmadd`/`fmsub`/`fnmadd`/`fnmsub` (fused multiply-add/substract) instructions.
- Adjust the translation of comparison operators, as immediate arguments are not supported.
- Adjust the translation of multiply-high operator (`Cmulhi`), as immediate arguments are not supported.

### Runtime glue

The runtime glue (`runtime/riscv.S`) consists of bits of assembly code used to switch between OCaml and C contexts. This includes calling external primitives, but also interacting with the GC and handling exceptions. This part of the runtime system cannot be written in C as it needs fiddle with the contents of individual registers.

The assembly code does not currently sport CFA directives (used to help stack unwinding by debuggers). I am planning to add this, but perhaps in a later PR.

### CODE REVIEW

File: README.adoc
Comment (gasche) [inline README.adoc:65]: You say "64 bits" here, but there appears to be some support for `riscv32` below. Is 32 bits also supported?

File: asmcomp/riscv/arch.ml
Comment (gasche) [inline asmcomp/riscv/arch.ml]: I'm not sure why you used decimal here and hexadecimal in `is_cheap_operation`, but 2048 is the most well-known power of two anyway.

File: asmcomp/riscv/emit.mlp
Comment (gasche) [inline asmcomp/riscv/emit.mlp:287]: Is there a reason why you are checking the type of the destination here, and not in any other case? (In particular not in the stack->reg float case.)

The whole pattern-matching is slightly tricky, but without this condition one can see that you covering all cases except `{loc = Unknown}, _ | _, {loc = Unknown}`, so the failure case can be refined. With this extra checking condition (which I guess gives emit-time errors for compiler bugs rather than runtime errors, which is nice) you are covering all nonfloat moves, and all float-float moves, whose locations are known (plus some invalid float-nonfloat moves).

**Comment by rwmjones** [comment]
We've been carrying nojb's patch in Fedora since Nov 2016 (https://src.fedoraproject.org/rpms/ocaml/c/3faf9753d1af7494bbce2777ca8bc296e761bf02).  This means that we not only make the OCaml compiler available in Fedora/RISC-V, but we also compile ~140 OCaml packages with it.  It's had occasional problems over the years but he's always come up with timely fixes when we've reported them.

The HiFive Unleashed isn't a particularly stable development platform - the two boards I have are used as Fedora builders and often need rebooting once a day.  However QEMU instances - while not as fast - are stable and easier to manage.  We hope to have a lot more hardware available from several sources this year.

One improvement I would like to see in this patch (which is *not* a showstopper), is better support for the Compressed extension, ie better choice of registers so that compressed instructions can be used.  In our experience with GCC, compressed instructions lead to significant reductions in code size.

File: README.adoc
Comment (gasche) [reply README.adoc:65]: There are many different RISCV ISA extensions, is it documented somewhere which set of extensions we are assuming?

File: asmcomp/riscv/emit.mlp
Comment (gasche) [inline asmcomp/riscv/emit.mlp]: If we had float-point temporary registers, we could just convert from the source to the temporary and store from there. Don't we have those?

**Review by gasche — COMMENTED** [review]
Thanks for this work!

I looked at the beginning of the patch and have some beginner questions.

File: README.adoc
Comment (nojb) [reply README.adoc:65]: No, 32 bits is not supported. However, the instruction sets for 32/64-bits are very similar and some of the basic definitions in `emit.mlp` were written with the prospect of supporting both at some point in the future. No problem getting rid of that switch if it becomes confusing. If memory serves, it only affects: 1) the definition of the word size, 2) the choice of load/store instruction.

As for the precise ISA variant supported, it is the 64-bit "General" variant `RV64G` (see https://en.wikipedia.org/wiki/RISC-V for the precise description). But I don't think it is documented anywhere. I will add a note about it in the README I guess.

File: asmcomp/riscv/emit.mlp
Comment (nojb) [reply asmcomp/riscv/emit.mlp:287]: Probably the extra checking condition was a result of tracking down a bug. A glance at the other backends show that they typically match only on the type of the source, so I suggest we do the same here.

Note that also we do not cover the case where both the source and the destination have `{loc = Stack _}`.

File: asmcomp/riscv/emit.mlp
Comment (nojb) [reply asmcomp/riscv/emit.mlp]: Currently all floating-point registers are made available to the register allocator. We could reserve one for "scratch" to use in this case, but is it worth it?

**Comment by xavierleroy** [comment]
I read the code and will submit a full review soon. In summary:
* The 64-bit part looks good to me.
* The 32-bit part has a few bugs.  Since it's not getting any testing (I'm afraid), I suggest to remove RV-32 support.
* Please add a `asmcomp/riscv/NOTES.md` file, like in the other ports, with references and URLs for the ISA spec and the ABI spec you used.

File: asmcomp/riscv/arch.ml
Comment (nojb) [reply asmcomp/riscv/arch.ml]: Thanks, I switched to hex here as well.

File: asmcomp/riscv/emit.mlp
Comment (xavierleroy) [reply asmcomp/riscv/emit.mlp]: This code doesn't work in RV-32 because `fmv.x.d` is not supported.  You can't save a 64-bit FP register in a 32-bit integer register, anyway.  

You don't need to reserve a FP register: you can just use a normal, allocatable register but declare it destroyed by the `Istore` instruction.  You can see this trick at work in the AMD64 port.  Basically, you would emit the following code
```
    `	fcvt.s.d	ft0, {emit_reg i.arg.(0)}\n`;
    `	fsw	ft0, {emit_int ofs}({emit_reg i.arg.(1)})\n`;
```
and add a case to `Proc.destroyed_at_oper`:
```
  | Iop(Istore(Single, _, _)) -> [| phys_reg 100 |]
```

File: asmcomp/riscv/emit.mlp
Comment (xavierleroy) [inline asmcomp/riscv/emit.mlp:58]: Just curious: any reason to use `'.'` instead of `'$'` as escape character?  I'm asking because I have plans to to use dots in asm symbols for other purposes.

File: asmcomp/riscv/emit.mlp
Comment (xavierleroy) [inline asmcomp/riscv/emit.mlp]: `jr` is enough, but `jalr` is not wrong either.

File: asmcomp/riscv/proc.ml
Comment (xavierleroy) [inline asmcomp/riscv/proc.ml:144]: For RV-32 you need to realign `ofs` to 8 before using it.

File: asmcomp/riscv/proc.ml
Comment (xavierleroy) [inline asmcomp/riscv/proc.ml:217]: Likewise, for RV-32, a realignment to 8 is missing.

File: asmcomp/riscv/proc.ml
Comment (xavierleroy) [inline asmcomp/riscv/proc.ml]: Unlike any other 32-bit ABI I know, the ELF psABI for RV-32 does not use aligned pairs of 32-bit registers to pass 64-bit integers: unaligned pairs are OK too.  For instance, a function `f(int32, int64)` will receive its first argument in a0 and its second argument in (a1, a2), not in (a2, a3) as in the code above.

It becomes ugly when the first 7 argument registers are already used and an int64 is passed.  Then, the ABI mandates that the low 32 bits are passed in register a7 and the high 32 bits in the first stack slot!

File: runtime/riscv.S
Comment (xavierleroy) [inline runtime/riscv.S:94]: For RV-32 the GC expects an array of 32-bit integers, so the offsets must be spaced by 4 instead of by 8.

File: runtime/riscv.S
Comment (xavierleroy) [inline runtime/riscv.S]: Why `0x10` in the store instruction and `16` here?

File: asmcomp/riscv/selection.ml
Comment (xavierleroy) [inline asmcomp/riscv/selection.ml:55]: Apparently, "multiply" doesn't support immediate operands any more than "multiply high", so please group the two cases Cmulhi and Cmuli together.

**Review by xavierleroy — CHANGES_REQUESTED** [review]
All right, here is the review.  Something needs to be done for the RV-32 code: either update or remove.  The other comments are suggestions.

File: asmcomp/riscv/emit.mlp
Comment (gasche) [reply asmcomp/riscv/emit.mlp]: After reading more code (I'm not familiar at all with the backend), my impression is that some other backends do this differently: they use a fixed register for the operation (here a floating-point register) and they indicate it in `destroyed_at_oper` for this operation. For example `arm` uses `s14` for single-floating-point load/stores in this way. If I understand correctly, this does not prevent it from being used by the register allocator, just not across this operation.

Comment (nojb) [reply asmcomp/riscv/emit.mlp]: Indeed! I had overlooked this obvious solution.

File: asmcomp/riscv/emit.mlp
Comment (nojb) [reply asmcomp/riscv/emit.mlp:58]: No reason, so I switched to `'$'`.

File: runtime/riscv.S
Comment (nojb) [reply runtime/riscv.S]: Fixed.

File: asmcomp/riscv/selection.ml
Comment (nojb) [reply asmcomp/riscv/selection.ml:55]: Of course; fixed.

**Comment by nojb** [comment]
> All right, here is the review. Something needs to be done for the RV-32 code: either update or remove.

Thanks for the quick review!

Indeed, the 32-bit code was just a vestige of initial efforts in this direction, but as I could never test it, it was half-broken. So I removed all such code from the patch.

I think all other points have been addressed as well.

**Comment by xavierleroy** [comment]
> One improvement I would like to see in this patch (which is not a showstopper), is better support for the Compressed extension, ie better choice of registers so that compressed instructions can be used. In our experience with GCC, compressed instructions lead to significant reductions in code size.

You (@rwjones) or @nojb could try to set `Proc.rotate_registers` to `false`.  This would steer the register allocator towards using a0..a5 and s2..s3 (the 8 registers compactly-encoded in the Compressed extension) in preference to other caller- or callee- save registers.  However, this may have a negative performance impact on superscalar  RISC-V implementations that don't perform register renaming.

Other aspects of the Compressed extension (e.g. two-address instructions) would need a lot more changes to the code generator to be usable.

**Comment by rwmjones** [comment]
I believe you also need to put `.option rvc` note into assembler output files.  From a quick grep of the code it doesn't seem to do that yet.  Edit: I mean, if you do this it will try to output a compressed instruction if the conditions are met, but you may still need to select registers as you described in order for C instructions to be applicable.

Of course there is also the issue of whether C vs no-C should be selectable at run time.  The SHAKTI chip is the only released RV64 chip that I know of which doesn't support C at all, but all commonly available current and future RV64 chips have the C extension.

File: asmcomp/riscv/emit.mlp
Comment (gasche) [reply asmcomp/riscv/emit.mlp]: But then, if I understand correctly, you could also use this `destroy_at_oper` approach for some uses of the "temporary registers" you have today. You use `tmp1` and `tmp2`, the first one is used in many places so it would be an invasive change, but `tmp2` is used only for `Lswitch` so this would be an easy way to free a register.

Comment (nojb) [reply asmcomp/riscv/emit.mlp]: I think you may be right.

For info, `t0` (`reg_tmp1`) is also used in the assembly stubs (`riscv.S`) as a temporary register. Initially also `t1` (`reg_tmp2`) was also used in the assembly stubs, so I didn't mind using it in the code generator as well, but since the switch to "domain state", the assembly stubs only need one temporary register, so indeed it looks like we may be able to reuse the other one for the register allocator.

I will look into it.

Comment (nojb) [reply asmcomp/riscv/emit.mlp]: Actually after checking the assembly stubs again, it seems `t1` is still used there, so I need to think if indeed the register can be freed for the register allocator.

Comment (nojb) [reply asmcomp/riscv/emit.mlp]: OK, I think it should be possible to reuse `t0` for the register allocator.

This register is not callee-save according to the standard convention, so assumed killed by any C call. In particular, it can be safely overwritten in the routines where it is currently used in the assembly stubs (`caml_callback_exn*_asm`).

I pushed a commit with this change. Thanks!

File: runtime/riscv.S
Comment (xavierleroy) [inline runtime/riscv.S]: I'm lost in the way you juggle t0, t1, t2 here, but I hope this was tested and passed.

**Review by xavierleroy — COMMENTED** [review]
Thank you for  the changes, which address my suggestions.

I'm not approving yet because I'd like to have a CI solution before we merge this code in trunk.

**Comment by xavierleroy** [comment]
This is shaping up nicely.  However, before we merge, we need to have CI for this code: either the HiFive board at Inria, or a QEMU virtual machine, or something based on cross-compilation + QEMU user emulation.  @shindere : the ball is in your court; could you please synchronize with @nojb ?

**Comment by shindere** [comment]
> @shindere : the ball is in your court; could you please
> synchronize with @nojb ?

The real machine does not seem stable enough so I guess we will need
Qemu for a stable solution. Do you share this point of view, @nojb?

If yes, we can talk by private e-mail to figure out how best to proceed.

**Comment by nojb** [comment]
> @shindere : the ball is in your court; could you please synchronize with @nojb ?
> The real machine does not seem stable enough so I guess we will need Qemu for a stable solution. Do you share this point of view, @nojb? If yes, we can talk by private e-mail to figure out how best to proceed.

Yes, I think so too. I'll send you an email to coordinate.

File: runtime/riscv.S
Comment (nojb) [reply runtime/riscv.S]: Yes, it was tested, and passed :)

Briefly, both `t0` and `t1` were reserved as temporaries before the change. In the assembly stubs, `TMP` is used as a general "scratch" register and `ARG_DOMAIN_STATE_PTR` is only used in a couple of specific places to pass the domain state pointer among the `caml_callback*_asm` routines and the code of `caml_start_program`. After the change, `t0` can be used by the register allocator and `t1` remains reserved. So I changed `TMP` to use `t1` (which remains generally reserved), and `ARG_DOMAIN_STATE_PTR` (which is only used in a couple of places) to `t0`.

It is possible this change was unnecessary, but it was easier for me to check the few uses of `ARG_DOMAIN_STATE_PTR` rather than the many uses of `TMP`.

**Comment by nojb** [comment]
This branch is being checked in INRIA's CI : https://ci.inria.fr/ocaml/job/Risc-V/30/

**Comment by nojb** [comment]
Pending final approval, I think this is ready to merge.

**Review by xavierleroy — APPROVED** [review]
Great job with the VM for CI !  Now I shall merge.

**Comment by nojb** [comment]
Thanks!

**Comment by shindere** [comment]
Nicolás Ojeda Bär (2020/04/24 07:08 -0700):
> Thanks!

@nojb will you update the Jenkins job so that it now uses trunk?

You can refer to how other jobs are configured and do the same, I think.

**Comment by nojb** [comment]
> @nojb will you update the Jenkins job so that it now uses trunk? You can refer to how other jobs are configured and do the same, I think.

It's already done :) Thanks for the help with the CI, by the way.

**Comment by shindere** [comment]
Nicolás Ojeda Bär (2020/04/24 09:19 -0700):
>
> > @nojb will you update the Jenkins job so that it now uses trunk? You can refer to how other jobs are configured and do the same, I think.
>
> It's already done :) Thanks for the help with the CI, by the way.

It was a pleasure to work with you @nojb

**Comment by rwmjones** [comment]
This is fantastic news.  I think it's great that nojb persisted with this work for around 3½ years and it's finally upstream.  Even though this is now targeted at 4.12, we will attempt to maintain the backport against 4.11 in the upcoming release of Fedora 33.

**Comment by damiendoligez** [comment]
@rwmjones As this is a very low-risk change, we could still consider cherry-picking to 4.11. Would that save you a significant amount of work?

**Comment by rwmjones** [comment]
TBH we already did the cherry pick ourselves:
https://pagure.io/fedora-ocaml/commits/fedora-33-4.11.0-dev2

So while it might save us a bit of time going forwards, it's not a big deal for us.  It could help Debian however, so it might be worth it for them.

**Comment by xavierleroy** [comment]
I am in favor of releasing this RISC-V port with 4.11.  It's new code that doesn't affect the existing ports. Plus, it will look great on the release announcement :-)

**Comment by avsm** [comment]
Having the code generator in 4.11.0 rather than a cherry pick would be amazing for MirageOS! We could add official downstream support much more easily as we base our ocaml-freestanding directly from OCaml releases at present. (cc @samoht)

**Comment by nojb** [comment]
> I am in favor of releasing this RISC-V port with 4.11. It's new code that doesn't affect the existing ports. Plus, it will look great on the release announcement :-)

Sorry, I missed this discussion! @Octachron has it been decided whether this PR will be included in 4.11?

**Comment by Octachron** [comment]
I am in favor of including it in 4.11.0, since there is no users that could be negatively affected.

**Comment by Octachron** [comment]
I have cherry-picked the PR to 4.11.0 (with #9516), and updated trunk Changes entry.

**Comment by nojb** [comment]
> I have cherry-picked the PR to 4.11.0 (with #9516), and updated trunk Changes entry.

Thanks!

**Pull request merged**
